sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/http/HttpPipelineCallState.java

```diff
@@ -76,12 +76,17 @@ public HttpRequest getHttpRequest() {
         return this.httpRequest;
     }
 
-    @SuppressWarnings("MethodDoesntCallSuperMethod")
-    @Override
-    public HttpPipelineCallState clone() {
-        HttpPipelineCallState cloned = new HttpPipelineCallState(this.pipeline, this.httpRequest);
-        cloned.currentPolicyIndex = this.currentPolicyIndex;
+    /**
+     * Copies the current state of the {@link HttpPipelineCallState}.
+     * <p>
+     * This method must be used when a re-request is made in the pipeline.
+     *
+     * @return A new instance of this pipeline call state.
+     */
+    public HttpPipelineCallState copy() {
+        HttpPipelineCallState copy = new HttpPipelineCallState(this.pipeline, this.httpRequest);
+        copy.currentPolicyIndex = this.currentPolicyIndex;
```

Comment:
Not sure what copying the policy index would mean here. How will this be used as the index would change when the policy is added to the pipeline?

(Not a question for this PR. Let's discuss offline)

Reply:
The index is copied as this is how we determine the next policy to apply in the request-response flow. This copy allows for the following:

Say we have the following pipeline:

Policy A -> Retry -> Policy B

Retry will make a copy of the state with `index = 1`.

So, it will effectively look like:

Retry -> Policy B

If another attempt needs to happen it'll restart from `Retry` rather than `Policy A`.